src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs

```diff
@@ -788,19 +781,67 @@ private Task WaitConnectionLockAsync([CallerMemberName] string memberName = null
         private class Subscription : IDisposable
         {
             private readonly InvocationHandler _handler;
-            private readonly List<InvocationHandler> _handlerList;
+            private readonly InvocationHandlerList _handlerList;
 
-            public Subscription(InvocationHandler handler, List<InvocationHandler> handlerList)
+            public Subscription(InvocationHandler handler, InvocationHandlerList handlerList)
             {
                 _handler = handler;
                 _handlerList = handlerList;
             }
 
             public void Dispose()
             {
-                lock (_handlerList)
+                _handlerList.Remove(_handler);
+            }
+        }
+
+        private class InvocationHandlerList
+        {
+            private readonly List<InvocationHandler> _invocationHandlers;
+            // A lazy cached copy of the handlers that doesn't change for thread safety. 
+            // Adding or removing a handler sets this to null.
+            private InvocationHandler[] _copiedHandlers;
+
+            internal InvocationHandlerList(InvocationHandler handler)
+            {
+                _invocationHandlers = new List<InvocationHandler>() { handler };
+            }
+
+            internal InvocationHandler[] GetHandlers()
+            {
+                var handlers = _copiedHandlers;
+                if (handlers == null)
+                {
+                    lock (_invocationHandlers)
+                    {
+                        // Check if the handlers are set, if not we'll copy them over.
+                        if (_copiedHandlers == null)
+                        {
+                            _copiedHandlers = _invocationHandlers.ToArray();
```

Comment:
Ahh, I just thought of something. If _invocationHandlers is empty there set _copiedHandlers to Array.Empty. (Does ToArray do that internally? idk)